### PR TITLE
Range-ify std.path.dirName()

### PR DESCRIPTION
**Pull request by WalterBright**
http://dlang.org/phobos/std_path.html#.dirName

100% unit test coverage

First use of `std.range.choose()` !!

### CODE REVIEW

File: std/path.d [hunk 1]
```diff
@@ -457,38 +459,75 @@ unittest
...
+    if (isRandomAccessRange!R && hasSlicing!R && hasLength!R && isSomeChar!(ElementType!R) ||
+        is(StringTypeOf!R))
 {
-    import std.conv : to;
-    if (path.empty) return to!(typeof(return))(".");
+    import std.range : choose, only;
 
-    auto p = rtrimDirSeparators(path);
-    if (p.empty) return path[0 .. 1];
+    alias C = ElementEncodingType!R;
```
Comment (andralex) [inline std/path.d]: why the empty line?

File: std/path.d [hunk 2]
```diff
@@ -457,38 +459,75 @@ unittest
...
 
-    version (Windows)
-    {
-        if (isUNC(p) && uncRootLength(p) == p.length)
-            return p;
-        if (p.length == 2 && isDriveSeparator(p[1]) && path.length > 2)
-            return path[0 .. 3];
-    }
+    bool dot;
```
Comment (andralex) [inline std/path.d]: many empty lines

File: std/path.d [hunk 3]
```diff
@@ -457,38 +459,75 @@ unittest
...
+                    goto Lret;
+                }
+                if (p.length == 2 && isDriveSeparator(p[1]) && path.length > 2)
+                {
+                    p = path[0 .. 3];
+                    goto Lret;
+                }
+            }
+
+            {
```
Comment (andralex) [inline std/path.d]: why the extra scope?

Comment (WalterBright) [inline std/path.d]: because of the goto around the variable initialization - dmd complains

File: std/path.d (hunk 1 above)
Comment (WalterBright) [inline std/path.d]: I like to set off declarations separate from code.

File: std/path.d (hunk 2 above)
Comment (WalterBright) [inline std/path.d]: thought it logically looked nicer

**Comment by andralex** [comment]
nicely done!

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**